Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/Networking/Forge/Networking/FacepunchP2PClient.cs

```diff
@@ -0,0 +1,430 @@
+
+				//Set the port
+				SetPort(clientPort);
+
+				int connectCounter = 0;
+				Task.Queue(() =>
+				{
+					do
+					{
+						// Send the accept headers to the server to validate
+						Client.Send(connectHeader, connectHeader.Length, hostId, P2PSend.Reliable);
+						Thread.Sleep(3000);
+					} while (!headerExchanged && IsBound && ++connectCounter < CONNECT_TRIES);
+
+					if (connectCounter >= CONNECT_TRIES)
+					{
+						if (connectAttemptFailed != null)
+							connectAttemptFailed(this);
+					}
+				});
+			}
+			catch (Exception e)
+			{
+				Logging.BMSLog.LogException(e);
+				// Do any generic initialization in result of the binding failure
+				OnBindFailure();
+
+				throw new FailedBindingException("Failed to bind to server's SteamId, see inner exception", e);
+			}
+		}
+
+		/// <summary>
+		/// Connects to a Steam Lobby then kicks off connection to the lobby owner's SteamId
+		/// </summary>
+		/// <param name="lobbyToJoin">The <see cref="Steamworks.Data.Lobby"/> to join</param>
+		/// <param name="pendCreates">Set the NetWorker::PendCreates to true</param>
+		private async void ConnectToLobbyAsync(Steamworks.Data.Lobby lobbyToJoin, bool pendCreates)
+		{
+			RoomEnter roomEnter = await lobbyToJoin.Join();
+			if (roomEnter != RoomEnter.Success)
+				return;
+			Lobby = lobbyToJoin;
+			Connect(Lobby.Owner.Id, pendCreates);
+		}
+
+		/// <summary>
+		/// Disconnect this client from the server
+		/// </summary>
+		/// <param name="forced">Used to tell if this disconnect was intentional <c>false</c> or caused by an exception <c>true</c></param>
+		public override void Disconnect(bool forced)
+		{
+			Logging.BMSLog.Log("<color=cyan>FacepunchP2P client disconnecting...</color>");
+
+			if (Lobby.Id.Value > 0)
+				Lobby.Leave();
+
+			if (Client == null)
+				return;
+
+			lock (Client)
+			{
+				if (forced)
+					CloseConnection();
+				else
+				{
+					var frame = new ConnectionClose(Time.Timestep, false, Receivers.Server, MessageGroupIds.DISCONNECT, false);
+					Send(frame, true);
+					Task.Queue(CloseConnection, 1000);
+				}
+
+				// Send signals to the methods registered to the disconnect events
+				if (forced)
+					OnForcedDisconnect();
+			}
+		}
+
+		/// <summary>
+		/// Infinite loop listening for new data from all connected clients on a separate thread.
+		/// This loop breaks when readThreadCancel is set to true
+		/// </summary>
+		private void ReadNetwork()
+		{
+			var messageFrom = default(SteamId);
+
+			try
+			{
+				BMSByte packet = null;
+				// Intentional infinite loop
+				while (IsBound)
+				{
+					// If the read has been flagged to be canceled then break from this loop
+					if (readThreadCancel)
+						return;
+
+					try
+					{
+						if (SteamNetworking.IsP2PPacketAvailable())
+						{
+							packet = Client.Receive(out messageFrom);
+						}
+						else
+						{
+							Thread.Sleep(1);
+							continue;
+						}
+
+						if (messageFrom == default(SteamId))
```

Comment:
Might be good to cache `default(SteamId)` in a class variable and just reuse that variable here?

Reply:
Good call, done.